packages/react-components/react-utilities/src/compose/resolveShorthand.ts

```diff
@@ -1,17 +1,16 @@
 import { isValidElement } from 'react';
 import type { ReplaceNullWithUndefined, SlotShorthandValue, UnknownSlotProps } from './types';
 
-export type ResolveShorthandOptions<Props, Required extends boolean = false> = {
-  required?: Required;
-  defaultProps?: Props;
-};
+export type ResolveShorthandOptions<Props> =
+  | { required: true; defaultProps?: Props }
+  | { required?: false; defaultProps?: Props };
 
 export type ResolveShorthandFunction<Props extends UnknownSlotProps = UnknownSlotProps> = {
   <P extends Props | null>(
     value: P | SlotShorthandValue | undefined,
-    options?: ResolveShorthandOptions<P, true>,
+    options: Extract<ResolveShorthandOptions<P>, { required: true }>,
```

Comment:
It took me a while to figure out what the purpose of this `Extract<>` is. As far as I can tell, it is selecting just the first part of the discriminated union of `ResolveShorthandOptions` right? I.e. this is effectively `{ required: true; defaultProps?: Props }`.

Perhaps this could be simplified to just have the options type inline. Since this whole thing is already super complex, anything that can be simplified helps.
```suggestion
    options: { required: true; defaultProps?: P },
```

Reply:
Sounds good to me also. I'll inline the types and deprecate `ResolveShorthandOptions`

Reply:
Just to avoid any kind of deprecations, I decided to go with the approach of making the condition internal to the `ResolveShorthandOptions` type, that way I don't need to deprecate and I can maintain the same behaviour

Reply:
If I were to go with a deprecation I'd deprecate both `ResolveShorthandOptions` and `ResolveShorthandFunction` and inline everything, but since we already got the types, let's leave as it is for now